pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,315 @@
+				active_issuance_post.saturating_add(passive_issuance),
+			),
+		};
+
+		match self {
+			Curve::PolynomialFunction(params) => params.calculate_costs(low, high),
+			Curve::SquareRootBondingFunction(params) => params.calculate_costs(low, high),
+			_ => todo!(),
+		}
+	}
+}
+
+pub trait BondingFunction<F: FixedSigned + PartialOrd> {
+	/// Calculates the cost of the curve between two points.
+	///
+	/// # Parameters
+	/// - `low`: The lower bound of the range for which the cost is to be calculated.
+	/// - `high`: The upper bound of the range for which the cost is to be calculated.
+	///
+	/// # Returns
+	/// - `Ok(F)`: The calculated cost if the operation is successful.
+	/// - `Err(ArithmeticError)`: An error if the calculation fails due to arithmetic issues.
+	///
+	/// # Errors
+	/// This function will return an `ArithmeticError` if the calculation cannot be performed.
+	fn calculate_costs(&self, low: F, high: F) -> Result<F, ArithmeticError>;
+}
+
+/// A struct representing the parameters of a polynomial function F(x) = mx³ + nx² + ox.
+///
+/// This struct is generic over the type `F`, which represents the type of the coefficients.
+///
+/// # Attributes
+/// - `m`: The coefficient for the quadratic term.
+/// - `n`: The coefficient for the linear term.
+/// - `o`: The constant term.
+///
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct PolynomialFunctionParameters<F> {
+	pub m: F,
+	pub n: F,
+	pub o: F,
+}
+
+impl<F> BondingFunction<F> for PolynomialFunctionParameters<F>
+where
+	F: FixedSigned,
+{
+	/// Calculates the cost for a polynomial bonding function within a given range.
+	///
+	/// The polynomial function is defined as `F(x) = m * x^3 + n * x^2 + o * x`.
+	/// This method calculates the difference `F(high) - F(low)` using a factored form to improve performance and reduce overflow risk.
+	///
+	/// The calculation follows these steps:
+	/// 1. Compute `delta_x = high - low`.
+	/// 2. Compute intermediate terms to reduce overflow risk:
+	///    - `high_plus_low = high + low`
+	///    - `high_low_mul = high * low`
+	///    - `high2 = high * high`
+	///    - `low2 = low * low`
+	/// 3. Compute the factorized cubic term: `cubic_term = high^2 + high * low + low^2`.
+	/// 4. Calculate the terms:
+	///    - `term1 = m * cubic_term`
+	///    - `term2 = n * high_plus_low`
+	/// 5. Sum the terms and multiply by `delta_x` to get the final result.
+	///
+	/// # Parameters
+	/// - `low`: The lower bound of the range.
+	/// - `high`: The upper bound of the range.
+	///
+	/// # Returns
+	/// - `Result<F, ArithmeticError>`: The calculated cost or an arithmetic error if an overflow or underflow occurs.
+	///
+	/// # Errors
+	/// - `ArithmeticError::Underflow`: If subtraction results in an underflow.
+	/// - `ArithmeticError::Overflow`: If any arithmetic operation results in an overflow.
+	fn calculate_costs(&self, low: F, high: F) -> Result<F, ArithmeticError> {
```

Comment:
If it is not obvious how the values are calculated, I can provide some additional information, how I transformed the terms.